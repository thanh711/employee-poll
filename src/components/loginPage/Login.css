.notice {
    color: red;
}

.custom-dropdown {
    appearance: none;
    -webkit-appearance: none;
    -moz-appearance: none;
    background-color: #f9f9f9;
    border: 1px solid #ccc;
    padding: 10px;
    font-size: 16px;
    border-radius: 5px;
    color: #333;
    width: 200px;
    cursor: pointer;
    transition: background-color 0.3s ease, border-color 0.3s ease;
}
  
.custom-dropdown::after {
    content: 'â–¼';
    font-size: 12px;
    color: #333;
    position: absolute;
    right: 15px;
    top: 50%;
    transform: translateY(-50%);
    pointer-events: none;
}
  
.custom-dropdown:hover {
    background-color: #f1f1f1;
    border-color: #888;
}
  
.custom-dropdown:focus {
    border-color: #4CAF50;
    box-shadow: 0 0 5px rgba(76, 175, 80, 0.5);
    outline: none;
}
  
.center-option {
    text-align: center;
}
h1 {
    font-family: Arial, sans-serif;
    color: #333;
}
   
.image-container {
    margin: 20px 0;
}
   
.login-image {
    width: 200px;
    height: 200px;
    border-radius: 50%;
}
   
.login-form {
    display: inline-block;
    text-align: left;
    max-width: 300px;
    width: 100%;
    margin: 0 auto;
}
   
  h2 {
    text-align: center;
    font-family: Arial, sans-serif;
}
   
.input-group {
    margin-bottom: 15px;
}
   
.input-field {
    width: 100%;
    padding: 10px;
    font-size: 16px;
    border: 1px solid #ccc;
    border-radius: 5px;
}
  