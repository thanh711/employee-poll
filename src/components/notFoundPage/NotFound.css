.not-found-container {
    text-align: center;
    padding: 50px;
    background-color: #f4f4f4;
    min-height: 100vh;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    color: #333;
  }
  
  .not-found-title {
    font-size: 10rem;
    margin: 0;
    color: #ff4040;
    text-shadow: 2px 2px 10px rgba(0, 0, 0, 0.1);
  }
  
  .not-found-message {
    font-size: 1.5rem;
    margin-top: 20px;
    margin-bottom: 40px;
  }
  
  .not-found-astronaut img {
    width: 300px;
    margin-bottom: 40px;
  }
  
  .go-home-button {
    padding: 10px 25px;
    font-size: 1.2rem;
    background-color: #ff4040;
    color: #fff;
    text-decoration: none;
    border-radius: 30px;
    box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
    transition: background-color 0.3s ease;
  }
  
  .go-home-button:hover {
    background-color: #ff2020;
  }
  