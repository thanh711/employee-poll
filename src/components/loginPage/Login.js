import "./Login.css";
import { useState, useEffect} from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate, useLocation } from "react-router-dom";
import Loading  from '../common/Loading';
import { handleGetAllQuestions } from '../../actions/questions';
import { handleGetAllUsers, loginByUser } from '../../actions/users';

const  Login= ()=> {
  const location = useLocation();
  const navigate = useNavigate();
  const dispatch = useDispatch(); 
  const users = useSelector(state => state.users);  
  const questions = useSelector(state => state.questions);  
  const [userId,setUserId]=useState('');
  const [password,setPassword]=useState('');


  useEffect(() => {
    if(users.length<=0)
    dispatch(handleGetAllUsers());
    if(questions.length<=0)
    dispatch(handleGetAllQuestions());
  }, [dispatch, users, questions]);

  const selectUser=(event)=>{
    const user=users.find(x=>x.id===event.target.value);
    if(user){
      setUserId(user.id);
      setPassword(user.password);
    }
  }

  const Login = ()=>{
    if(userId&&password){
      const currentUser=users.find(x=>x.id===userId);
      if(currentUser&&currentUser.password===password){
        dispatch(loginByUser(currentUser))
        if (location.state?.previousPage) 
          navigate(location.state?.previousPage)
        else 
          navigate("/");
      }else{
        alert('username or password incorrect please try again !!!');
      }
    }else{
      alert('please enter username and password and try again !!!');
    }
  }
  
  return (
    <div className="container">
     {users.length>0?(<><h1>Employee Polls </h1>
      <div className="image-container">
        <img
          src='/img/login.webp '
          alt="Employee Polls"
          className="login-image"
        />
      </div>
      <h2>Log In</h2>
      <select className="custom-dropdown" onChange={selectUser} defaultValue="none">
        <option value="none" disabled className="center-option">
          -- Select an option --
        </option>
        {users.map((user) => (
          <option key={user.id} value={user.id} >
            {user.id}
          </option>
        ))}
      </select>
      <br></br>
      <br></br>
        <div className="input-group">
          <input 
          type="text" 
          placeholder="User" 
          className="input-field"  
          value={userId}
          onChange={(e)=>setUserId(e.target.value)}/>
        </div>
        <div className="input-group">
          <input 
          type="password" 
          placeholder="Password" 
          className="input-field" 
          value={password}
          onChange={(e)=>setPassword(e.target.value)} />
        </div>
        <button type="submit" className="submit-btn" onClick={Login}>
          Login
        </button>
        </>) :(<Loading/>)}
    </div>
  );
}
 
export default Login;