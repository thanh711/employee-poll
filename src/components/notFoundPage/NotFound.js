import React from 'react';
import './NotFound.css'; 
import { Link } from "react-router-dom";

const NotFound = () => {
  return (
    <div className="not-found-container">
      <h1 className="not-found-title">404</h1>
      <p className="not-found-message">
        Oops! It seems like you're lost in space. <br />
        The page you're looking for is out of this world!
      </p>
      <div className="not-found-astronaut">
        <img 
          src="/img/notfound.webp" 
          alt="not found" 
        />
      </div>
      <Link to={'/'}  className='go-home-button'>Take Me Home</Link>
    </div>
  );
};

export default NotFound;
