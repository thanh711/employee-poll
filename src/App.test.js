import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from "redux-mock-store"
import { BrowserRouter } from 'react-router-dom';
import { render, screen,  fireEvent } from '@testing-library/react';
import NotFound from "./components/notFoundPage/NotFound";
import Login from "./components/loginPage/Login";

var _DATA = require('./_DATA');

describe("_saveQuestion() test", () => {
  it("_saveQuestion() success", async () => {
    const optionOneText = "Tea";
    const optionTwoText = "Coffee";
    const author = "thanhnc34";
    const question = {
      optionOneText,
      optionTwoText,
      author,
    };
    var result = await _DATA._saveQuestion(question);
    expect(result).toBeDefined();
    expect(result).toHaveProperty('id');
    expect(result).toHaveProperty('timestamp');
    expect(result).toHaveProperty('author', author);
    expect(result.optionOne).toHaveProperty('text', optionOneText);
    expect(result.optionTwo).toHaveProperty('text',optionTwoText);
    expect(result.optionTwo).toHaveProperty('votes', []);
    expect(result.optionOne).toHaveProperty('votes', []);
  });

  it("_saveQuestion() fail", async () => {
    const question = {};
    expect(_DATA._saveQuestion(question)).rejects.toEqual(
      "Please provide optionOneText, optionTwoText, and author"
    );
  });
});

describe("_saveQuestionAnswer() test", () => {
  it("_saveQuestionAnswer() success", async () => {
    const authedUser = "tylermcginnis";
    const qid = "vthrdm985a262al8qx3do";
    const answer = "optionOne";
    const question = {
      authedUser,
      qid,
      answer,
    };
    var result = await _DATA._saveQuestionAnswer(question);
    expect(result).toBeDefined();
    expect(result).toBe(true);
  });

  it("_saveQuestionAnswer() fail", async () => {
    const question = {};
    expect(_DATA._saveQuestionAnswer(question)).rejects.toEqual(
      "Please provide authedUser, qid, and answer"
    );
  });
});

describe('NotFound component', () => {
  it('matches the snapshot', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <NotFound/>
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Login component', () => {
  beforeAll(()=>{
    jest.spyOn(window,"alert").mockImplementation(()=>{});
  });
  afterAll(()=>{
    window.alert.mockRestore();
  });
  const mockStore=configureMockStore();
  const initialState={
    users:[{
      id: 'sarahedo',
      password:'password123',
      name: 'Sarah Edo',
      avatarURL: '/img/sarahedo.webp',
      answers: {
        "8xf0y6ziyjabvozdd253nd": 'optionOne',
        "6ni6ok3ym7mf1p33lnez": 'optionOne',
        "am8ehyc8byjqgar0jgpub9": 'optionTwo',
        "loxhs1bqm25b708cmbf3g": 'optionTwo'
      },
      questions: ['8xf0y6ziyjabvozdd253nd', 'am8ehyc8byjqgar0jgpub9']
    }],
    questions:[{
      id: 'am8ehyc8byjqgar0jgpub9',
      author: 'sarahedo',
      timestamp: 1488579767190,
      optionOne: {
        votes: [],
        text: 'conduct a release retrospective 1 week after a release',
      },
      optionTwo: {
        votes: ['sarahedo'],
        text: 'conduct release retrospectives quarterly'
      }
    }]
  }
  const store= mockStore(initialState)


  it('should alert when click login button without username and password', () => {
    render(
    <Provider store={store}>
      <BrowserRouter>
        <Login />
      </BrowserRouter>
    </Provider>
  );

  const submitButton = screen.getByRole('button', { name: /login/i });
  fireEvent.click(submitButton)
  expect(window.alert).toHaveBeenCalledWith("please enter username and password and try again !!!");
  });

  it('should render username field, password field, and submit button', () => {
     render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );

    const usernameField = screen.getByPlaceholderText('User');
    const passwordField = screen.getByPlaceholderText('Password');
    const submitButton = screen.getByRole('button', { name: /login/i });

    expect(usernameField).toBeInTheDocument();
    expect(passwordField).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();
  });

  it('should alert when click login button with incorrect username and password', () => {
    render(
     <Provider store={store}>
       <BrowserRouter>
         <Login />
       </BrowserRouter>
     </Provider>
   );

   const usernameField = screen.getByPlaceholderText('User');
   const passwordField = screen.getByPlaceholderText('Password');
   const submitButton = screen.getByRole('button', { name: /login/i });
   fireEvent.change(usernameField,{target:{value:"sarahedo"}})
   fireEvent.change(passwordField,{target:{value:"incorrect password"}})
   fireEvent.click(submitButton)
   expect(window.alert).toHaveBeenCalledWith("username or password incorrect please try again !!!");
  });

  it(`should navigte to home page at route "/" when login success`, () => {
    render(
     <Provider store={store}>
       <BrowserRouter>
         <Login />
       </BrowserRouter>
     </Provider>
   );
   
   const usernameField = screen.getByPlaceholderText('User');
   const passwordField = screen.getByPlaceholderText('Password');
   const submitButton = screen.getByRole('button', { name: /login/i });
   fireEvent.change(usernameField,{target:{value:"sarahedo"}})
   fireEvent.change(passwordField,{target:{value:"password123"}})
   fireEvent.click(submitButton);
   expect(window.location.pathname).toBe('/');
  });

  it('matches the snapshot', () => {
    const { asFragment } =  render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

