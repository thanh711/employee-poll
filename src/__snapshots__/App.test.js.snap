// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h1>
      Employee Polls 
    </h1>
    <div
      class="image-container"
    >
      <img
        alt="Employee Polls"
        class="login-image"
        src="/img/login.webp "
      />
    </div>
    <h2>
      Log In
    </h2>
    <select
      class="custom-dropdown"
    >
      <option
        class="center-option"
        disabled=""
        selected=""
        value="none"
      >
        -- Select an option --
      </option>
      <option
        value="sarahedo"
      >
        sarahedo
      </option>
    </select>
    <br />
    <br />
    <div
      class="input-group"
    >
      <input
        class="input-field"
        placeholder="User"
        type="text"
        value=""
      />
    </div>
    <div
      class="input-group"
    >
      <input
        class="input-field"
        placeholder="Password"
        type="password"
        value=""
      />
    </div>
    <button
      class="submit-btn"
      type="submit"
    >
      Login
    </button>
  </div>
</DocumentFragment>
`;

exports[`NotFound component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="not-found-container"
  >
    <h1
      class="not-found-title"
    >
      404
    </h1>
    <p
      class="not-found-message"
    >
      Oops! It seems like you're lost in space. 
      <br />
      The page you're looking for is out of this world!
    </p>
    <div
      class="not-found-astronaut"
    >
      <img
        alt="not found"
        src="/img/notfound.webp"
      />
    </div>
    <a
      class="go-home-button"
      href="/"
    >
      Take Me Home
    </a>
  </div>
</DocumentFragment>
`;
